<h1>Script-mode Custom Training Container</h1>

This notebook demonstrates how to build and use a custom Docker container for training with Amazon SageMaker that leverages on the <strong>Script Mode</strong> execution that is implemented by the sagemaker-containers library.

We start by defining some variables like the current execution role, the ECR repository that we are going to use for pushing the custom Docker container and a default Amazon S3 bucket to be used by Amazon SageMaker.

In [2]:
import boto3
import sagemaker
from sagemaker import get_execution_role

ecr_namespace = 'sagemaker-training-containers/'
prefix = 'script-mode-container'

ecr_repository_name = ecr_namespace + prefix
role = get_execution_role()

account_id = role.split(':')[4]
region = boto3.Session().region_name
sagemaker_session = sagemaker.session.Session()
bucket = sagemaker_session.default_bucket()

print(account_id)
print(region)
print(role)
print(bucket)

476245144988
ap-southeast-1
arn:aws:iam::476245144988:role/service-role/AmazonSageMaker-ExecutionRole-20191204T094674
sagemaker-ap-southeast-1-476245144988


Let's take a look at the Dockerfile which defines the statements for building our script-mode custom training container:

In [3]:
! pygmentize ../docker/Dockerfile

# Part of the implementation of this container is based on the Amazon SageMaker Apache MXNet container.
# https://github.com/aws/sagemaker-mxnet-container

FROM ubuntu:16.04

LABEL maintainer="Giuseppe A. Porcelli"

# Defining some variables used at build time to install Python3
ARG PYTHON=python3
ARG PYTHON_PIP=python3-pip
ARG PIP=pip3
ARG PYTHON_VERSION=3.6.6

# Install some handful libraries like curl, wget, git, build-essential, zlib
RUN apt-get update && apt-get install -y --no-install-recommends software-properties-common && \
    add-apt-repository ppa:deadsnakes/ppa -y && \
    apt-get update && apt-get install -y --no-install-recommends \
        build-essential \
        ca-certificates \
        curl \
        wget \
        git \
        libopencv-dev \
        openssh-client \
        openssh-server \
        vim \
        zlib1g-dev && \
    rm -rf /var/lib/apt/lists/*

# Installing Python3
RUN wget https://www.python.org/ftp/python/$PYTHON_VERSION/Python-$PYTHON_VERSION.

At high-level the Dockerfile specifies the following operations for building this container:
<ul>
    <li>Start from Ubuntu 16.04</li>
    <li>Define some variables to be used at build time to install Python 3</li>
    <li>Some handful libraries are installed with apt-get</li>
    <li>We then install Python 3 and create a symbolic link</li>
    <li>We install some Python libraries like numpy, pandas, ScikitLearn, etc.</li>
    <li>We set e few environment variables, including PYTHONUNBUFFERED which is used to avoid buffering Python standard output (useful for logging)</li>
    <li>We install the <strong>sagemaker-containers</strong> library</li>
    <li>Finally, we copy all contents in <strong>code/</strong> (which is where our training code is) under <strong>/opt/ml/code/</strong> which is the path where sagemaker-containers expect to find training code</li>
</ul>

<h3>Build and push the container</h3>
We are now ready to build this container and push it to Amazon ECR. This task is executed using a shell script stored in the ../script/ folder. Let's take a look at this script and then execute it.

In [4]:
! pygmentize ../scripts/build_and_push.sh

ACCOUNT_ID=$1
REGION=$2
REPO_NAME=$3

docker build -f ../docker/Dockerfile -t $REPO_NAME ../docker

docker tag $REPO_NAME $ACCOUNT_ID.dkr.ecr.$REGION.amazonaws.com/$REPO_NAME:latest

$(aws ecr get-login --no-include-email --registry-ids $ACCOUNT_ID)

aws ecr describe-repositories --repository-names $REPO_NAME || aws ecr create-repository --repository-name $REPO_NAME

docker push $ACCOUNT_ID.dkr.ecr.$REGION.amazonaws.com/$REPO_NAME:latest


<h3>--------------------------------------------------------------------------------------------------------------------</h3>

The script builds the Docker container, then creates the repository if it does not exist, and finally pushes the container to the ECR repository. The build task requires a few minutes to be executed the first time, then Docker caches build outputs to be reused for the subsequent build operations.

In [5]:
! ../scripts/build_and_push.sh $account_id $region $ecr_repository_name

Sending build context to Docker daemon  11.26kB
Step 1/16 : FROM ubuntu:16.04
 ---> 4b22027ede29
Step 2/16 : LABEL maintainer="Giuseppe A. Porcelli"
 ---> Using cache
 ---> d2c3773b03a9
Step 3/16 : ARG PYTHON=python3
 ---> Using cache
 ---> 43b860d1aa35
Step 4/16 : ARG PYTHON_PIP=python3-pip
 ---> Using cache
 ---> f291d54a9e49
Step 5/16 : ARG PIP=pip3
 ---> Using cache
 ---> ad2298b40cad
Step 6/16 : ARG PYTHON_VERSION=3.6.6
 ---> Using cache
 ---> d737b6383031
Step 7/16 : RUN apt-get update && apt-get install -y --no-install-recommends software-properties-common &&     add-apt-repository ppa:deadsnakes/ppa -y &&     apt-get update && apt-get install -y --no-install-recommends         build-essential         ca-certificates         curl         wget         git         libopencv-dev         openssh-client         openssh-server         vim         zlib1g-dev &&     rm -rf /var/lib/apt/lists/*
 ---> Using cache
 ---> 832ff837a8db
Step 8/16 : RUN wget https://www.python.org/ftp/python/$P

<h3>Training with Amazon SageMaker</h3>

Once we have correctly pushed our container to Amazon ECR, we are ready to start training with Amazon SageMaker, which requires the ECR path to the Docker container used for training as parameter for starting a training job.

In [6]:
container_image_uri = '{0}.dkr.ecr.{1}.amazonaws.com/{2}:latest'.format(account_id, region, ecr_repository_name)
print(container_image_uri)

476245144988.dkr.ecr.ap-southeast-1.amazonaws.com/sagemaker-training-containers/script-mode-container:latest


Given the purpose of this example is explaining how to build custom script-mode containers, we are not going to train a real model. The script that will be executed does not define a specific training logic; it just outputs the configurations injected by SageMaker and implements a dummy training loop. Training data is also dummy. Let's analyze the script first:

In [7]:
! pygmentize ../docker/code/train.py

from __future__ import absolute_import

import sys
import time
import os
import argparse

from utils import save_model_artifacts, print_files_in_path

def train(hp1, hp2, hp3, train_channel, validation_channel):

    print('\nList of files in train channel: ')
    print_files_in_path(os.environ['SM_CHANNEL_TRAIN'])
    
    print('\nList of files in validation channel: ')
    print_files_in_path(os.environ['SM_CHANNEL_VALIDATION'])
    
    # Dummy net.
    net = None
        
    # Run training loop.
    epochs = 5
    for x in range(epochs):
        print("\nRunning epoch {0}...".format(x))

        time.sleep(30)

        print("Completed epoch {0}.".format(x))
        
    # At the end of the training loop, we have to save model artifacts.
    model_dir = os.environ['SM_MODEL_DIR']
    save_model_artifacts(model_dir + '/', net)

if __name__ == "__main__":

    parser = argparse.ArgumentParser()
    
    # sagemaker-containers passes hyperparameters as arguments
    parser.add_argum

You can realize that the training code has been implemented as a standard Python script, that will be invoked by the sagemaker-containers library passing hyperparameters as arguments. This way of invoking training script is indeed called <strong>Script Mode</strong> for Amazon SageMaker containers.

Now, we upload some dummy data to Amazon S3, in order to define our S3-based training channels.

In [8]:
! echo "val1, val2, val3" > dummy.csv
print(sagemaker_session.upload_data('dummy.csv', bucket, prefix + '/train'))
print(sagemaker_session.upload_data('dummy.csv', bucket, prefix + '/val'))
! rm dummy.csv

s3://sagemaker-ap-southeast-1-476245144988/script-mode-container/train/dummy.csv
s3://sagemaker-ap-southeast-1-476245144988/script-mode-container/val/dummy.csv


### Local Mode

Finally, we can execute the training job by calling the fit() method.

In [9]:
import sagemaker
import json

# JSON encode hyperparameters to avoid showing some info messages raised by the sagemaker-containers library.
def json_encode_hyperparameters(hyperparameters):
    return {str(k): json.dumps(v) for (k, v) in hyperparameters.items()}

hyperparameters = json_encode_hyperparameters({
    "hp1": "value1",
    "hp2": 300,
    "hp3": 0.001})

est = sagemaker.estimator.Estimator(container_image_uri,
                                    role, 
                                    train_instance_count=1, 
                                    train_instance_type='local', # we use local mode
                                    #train_instance_type='ml.m5.xlarge',
                                    base_job_name=prefix,
                                    hyperparameters=hyperparameters)

train_config = sagemaker.session.s3_input('s3://{0}/{1}/train/'.format(bucket, prefix), content_type='text/csv')
val_config = sagemaker.session.s3_input('s3://{0}/{1}/val/'.format(bucket, prefix), content_type='text/csv')

est.fit({'train': train_config, 'validation': val_config })

Parameter image_name will be renamed to image_uri in SageMaker Python SDK v2.
's3_input' class will be renamed to 'TrainingInput' in SageMaker Python SDK v2.
's3_input' class will be renamed to 'TrainingInput' in SageMaker Python SDK v2.


Creating tmpzsa45gko_algo-1-rdfp4_1 ... 
Attaching to tmpzsa45gko_algo-1-rdfp4_12mdone
algo-1-rdfp4_1  | 2020-08-31 10:09:55,429 sagemaker-containers INFO     No GPUs detected (normal if no gpus installed)
algo-1-rdfp4_1  | 2020-08-31 10:09:55,452 sagemaker-containers INFO     No GPUs detected (normal if no gpus installed)
algo-1-rdfp4_1  | 2020-08-31 10:09:55,469 sagemaker-containers INFO     No GPUs detected (normal if no gpus installed)
algo-1-rdfp4_1  | 2020-08-31 10:09:55,483 sagemaker-containers INFO     Invoking user script
algo-1-rdfp4_1  | 
algo-1-rdfp4_1  | Training Env:
algo-1-rdfp4_1  | 
algo-1-rdfp4_1  | {
algo-1-rdfp4_1  |     "additional_framework_parameters": {},
algo-1-rdfp4_1  |     "channel_input_dirs": {
algo-1-rdfp4_1  |         "train": "/opt/ml/input/data/train",
algo-1-rdfp4_1  |         "validation": "/opt/ml/input/data/validation"
algo-1-rdfp4_1  |     },
algo-1-rdfp4_1  |     "current_host": "algo-1-rdfp4",
algo-1-rdfp4_1  |     "framework_module": null,
algo